src/main/java/org/truffleruby/language/loader/RequireNode.java

```diff
@@ -156,34 +158,65 @@ private DynamicObject mimeTypeNotFound(String expandedPath, String mimeType) {
     private void requireCExtension(String feature, String expandedPath) {
         final FeatureLoader featureLoader = getContext().getFeatureLoader();
 
+        final List<TruffleObject> libraries;
+
         try {
             featureLoader.ensureCExtImplementationLoaded(feature, this);
 
             if (getContext().getOptions().CEXTS_LOG_LOAD) {
                 Log.LOGGER.info(String.format("loading cext module %s (requested as %s)", expandedPath, feature));
             }
 
-            featureLoader.loadCExtLibrary(expandedPath);
+            libraries = featureLoader.loadCExtLibrary(expandedPath);
         } catch (Exception e) {
             handleCExtensionException(feature, e);
             throw e;
         }
 
-        final TruffleObject initFunction = getInitFunction(expandedPath);
+        final String initFunctionName = "Init_" + getBaseName(expandedPath);
+
+        final TruffleObject initFunction = findFunctionInLibraries(libraries, initFunctionName, expandedPath);
 
         if (!ForeignAccess.sendIsExecutable(isExecutableNode, initFunction)) {
-            CompilerDirectives.transferToInterpreterAndInvalidate();
-            throw new UnsupportedOperationException(initFunction + "() is not IS_EXECUTABLE");
+            throw new RaiseException(coreExceptions().loadError(initFunctionName + "() is not executable", expandedPath, null));
         }
 
         try {
             ForeignAccess.sendExecute(executeNode, initFunction);
         } catch (InteropException e) {
-            CompilerDirectives.transferToInterpreterAndInvalidate();
             throw new JavaException(e);
         }
     }
 
+    public TruffleObject findFunctionInLibraries(List<TruffleObject> libraries, String functionName, String path) {
+        Object initObject = null;
+
+        for (TruffleObject library : libraries) {
+            try {
+                initObject = ForeignAccess.sendRead(readNode, library, functionName);
+            } catch (UnknownIdentifierException e) {
+                // TODO CS 18-Mar-18 it's not ideal that we're catching and throwing an exception when we don't find Init_ in each file
+                continue;
+            } catch (UnsupportedMessageException e) {
+                continue;
+            }
```

Comment:
UnsupportedMessageException sounds like Sulong would be completely broken, so I think it'd be OK to just rethrow as a JavaException in that case.